pkg/api/customization/pipeline/pipeline_execution_log.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func onError(rw http.ResponseWriter, _ *http.Request, code int, err error) {
+	rw.WriteHeader(code)
+	rw.Write([]byte(err.Error()))
+}
+
+func (h *ExecutionHandler) handleLog(apiContext *types.APIContext) error {
+	stageInput := apiContext.Request.URL.Query().Get("stage")
+	stepInput := apiContext.Request.URL.Query().Get("step")
+	stage, err := strconv.Atoi(stageInput)
+	if err != nil {
+		return err
+	}
+	step, err := strconv.Atoi(stepInput)
+	if err != nil {
+		return err
+	}
+	ns, name := ref.Parse(apiContext.ID)
+	execution, err := h.PipelineExecutionLister.Get(ns, name)
+	if err != nil {
+		return err
+	}
+	clusterName, _ := ref.Parse(execution.Spec.ProjectName)
+	userContext, err := h.ClusterManager.UserContext(clusterName)
+	if err != nil {
+		return err
+	}
+
+	pipelineEngine := engine.New(userContext)
+
+	c, err := upgrader.Upgrade(apiContext.Response, apiContext.Request, nil)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	cancelCtx, cancel := context.WithCancel(apiContext.Request.Context())
+	readerGroup, ctx := errgroup.WithContext(cancelCtx)
+	apiContext.Request = apiContext.Request.WithContext(ctx)
+
+	go func() {
+		for {
+			if _, _, err := c.NextReader(); err != nil {
+				cancel()
+				c.Close()
+				break
+			}
+		}
+	}()
+
+	go func() {
+		readerGroup.Wait()
+	}()
+
+	syncT := time.NewTicker(logSyncInterval)
+	defer syncT.Stop()
+
+	prevLog := ""
+	for {
+		select {
+		case <-syncT.C:
+			execution, err = h.PipelineExecutionLister.Get(ns, name)
+			if err != nil {
+				logrus.Debugf("error in execution get: %v", err)
+				if prevLog == "" {
+					writeData(c, []byte("Log is unavailable."))
+				}
+				c.WriteControl(websocket.CloseMessage, []byte{}, time.Now().Add(writeWait))
+				return nil
+			}
+			log, err := pipelineEngine.GetStepLog(execution, stage, step)
+			if err != nil {
+				logrus.Debug(err)
+				if prevLog == "" {
+					writeData(c, []byte("Log is unavailable."))
+				}
+				c.WriteControl(websocket.CloseMessage, []byte{}, time.Now().Add(writeWait))
+				return nil
+			}
+			newLog := getNewLog(prevLog, log)
+			prevLog = log
+			if newLog != "" {
+				if err := writeData(c, []byte(newLog)); err != nil {
+					logrus.Debugf("error in writeData: %v", err)
+					return nil
+				}
+			}
+			if execution.Status.Stages[stage].Steps[step].Ended != "" {
+				c.WriteControl(websocket.CloseMessage, []byte{}, time.Now().Add(writeWait))
+				return nil
+			}
+		}
+	}
+}
+
+func writeData(c *websocket.Conn, buf []byte) error {
+	messageWriter, err := c.NextWriter(websocket.TextMessage)
+	if err != nil {
+		return err
+	}
+
+	if _, err := messageWriter.Write(buf); err != nil {
+		return err
+	}
+
+	return messageWriter.Close()
+}
+
+func getNewLog(prevLog string, currLog string) string {
+	if len(prevLog) < Threshold {
+		return strings.TrimPrefix(currLog, prevLog)
+	}
+	//long logs are trimmed so we use previous log tail to do comparison
+	prevLogTail := prevLog[len(prevLog)-Threshold:]
```

Comment:
Can we use prevLog[threadhold:] instead of prevLog[len(prevLog)-Threshold:] or just last 20 line?

Reply:
The context here is that when logs are long, what we get from Jenkins is trimed from the head, so we need to get the tail of prev log and try to find the increment log. The `prevLog[threadhold:]` may get a minimum of one charactor when length of the log is `threadhold+1` so is not proper for the check. But we can optimise to use a smaller tail for comparison.